validate: add shfmt

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

* The first commit: Added [shfmt](https://github.com/mvdan/sh) to `hack/validate`.
* The second commit: ran the following command:
```console
$ git grep --name-only '^#!' | egrep -v '(vendor|\.go|Jenkinsfile)' | 
  xargs shfmt -w -bn -ci -sr
```

Close https://github.com/moby/moby/issues/38055

**- How I did it**

Used the latest shfmt.

**- How to verify it**
Observe CI

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
:penguin:

AkihiroSuda committed:
validate: add shfmt

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
@tianon @thaJeztah @cpuguy83 PTAL

thaJeztah commented on hack/validate/shfmt:
```diff
@@ -0,0 +1,7 @@
+# NOTE: `git grep '^#!'` may also pick up non-shell script files.
+# Add `| grep -v exceptional_file.foo` if any false-positive is detected.
+git grep --name-only '^#!' | grep -v vendor | grep -v '\.go$' | grep -v Jenkinsfile |
+	xargs shfmt -d
```
I pushed a branch with an extra commit so see what it would look like; https://github.com/moby/moby/compare/master...thaJeztah:shfmt2_alt?expand=1&w=1

Or (easier to read) with indentation whitespaces ignored;
https://github.com/moby/moby/compare/master...thaJeztah:shfmt2_alt?expand=1&w=1

AkihiroSuda replied:
Both URLs seem same 😅

thaJeztah replied:
oh! my bad; one of them should not have `&w=1`

tianon replied:
Diff in https://github.com/moby/moby/compare/master...thaJeztah:shfmt2_alt?w=1 LGTM!
(thanks to you both for working on this :heart: :+1:)

AkihiroSuda replied:
updated to use the suggested flags, thanks

thaJeztah reviewed (changes requested):
Thanks! I like this, as it indeed would help us to catch incorrect formatting. I do think we should tweak the configuration a bit to match our current style (left suggestions inline, and pushed an example branch to see the result).

I'm also wondering if we should add a `.editorconfig` file in this repository, so that contributors editing the code would get the same settings as we use in our CI (if their IDE/editor supports `.editorconfig` files).

mvdan commented:
Just wanted to briefly mention that the tool supports EditorConfig natively too: https://github.com/mvdan/sh#shfmt

That might be a simpler and more consistent way to configure its options.

thaJeztah commented:
@mvdan yes, I saw that (mentioned it in a comment above); I think it's worth considering to do that

AkihiroSuda committed:
run shfmt

git grep --name-only '^#!' | egrep -v '(vendor|\.go|Jenkinsfile)' | xargs shfmt -w -bn -ci -sr

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

tianon commented on hack/make.sh:
```diff
@@ -34,18 +34,18 @@ echo
 	binary-daemon
 	dynbinary
-
+	\
```
This looks like a weird edge case that maybe a future version of `shfmt` can handle better (blank lines inside an array). 😅

AkihiroSuda replied:
removed

AkihiroSuda commented:
Most shell scripts in this repo do not have `*.sh` suffix. So probably EditorConfig isn't useful for this repo?

AkihiroSuda committed:
hack/make.sh: remove extra empty lines

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

tianon reviewed (approved):
This LGTM.

IMO we should land this (as it's a drastic improvement to consistency) and discuss `.editorconfig` and further tweaks separately (assuming CI ends up green, but I don't see any reason it wouldn't). 👍

AkihiroSuda commented on hack/make.sh:
```diff
@@ -34,18 +34,18 @@ echo
 	test-integration
 	test-docker-py
-
+	\
```
removed

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
> IMO we should land this (as it's a drastic improvement to consistency) and discuss `.editorconfig` and further tweaks separately (assuming CI ends up green, but I don't see any reason it wouldn't).

Yes, seems fine for a follow-up; I know there's been some discussion around "IDE-specific configuration" in the repo (although `.editorconfig` _should be_ quite universal).


Also, thanks @mvdan for this tool ❤️ I must admit I hadn't come round to testing it after it was last discussed, and I'm happy to see it does a great job 👍

Let me open a ticket for @tianon's suggestion https://github.com/moby/moby/pull/40607#discussion_r386782203

AkihiroSuda commented:
`All nodes of label ‘amd64&&ubuntu-1804&&overlay2’ are offline` :cry:

thaJeztah commented:
Hm.. let me ask @StefanScherer
